[DS-3971] BitstreamStorageServiceImpl: Do not call update

AlexanderS opened the pull request:
If the BitstreamStorageService calls update on the just cloned bitstream, this
bitstream will not belong to any Item and the user may get authorization
failures.

The update call can be removed here, because the clone method is only used by
the AbstractVersionProvider and it will call update on the bitstream later by
itself.

https://jira.duraspace.org/browse/DS-3971

Commit:
BitstreamStorageServiceImpl: Do not call update

If the BitstreamStorageService calls update on the just cloned bitstream, this
bitstream will not belong to any Item and the user may get authorization
failures.

The update call can be removed here, because the clone method is only used by
the AbstractVersionProvider and it will call update on the bitstream later by
itself.

mwoodiupui commented:
I dislike having BitstreamStorageServiceImpl make changes to an entity and then give up control of it without persisting them, but we may have no choice here.  If so, the unpersisted state should be documented.  The place for that is in BitstreamStorageService#clone, which is not documented at all.

Commit:
BitstreamStorageService: Add update() call comment

Add a comment, that BitstreamStorageService#clone does not call update because
of permission problems and that it is the callers responsibility to persist the
changes.

AlexanderS commented:
@mwoodiupui I just added some comment about the missing update() call to the existing comment of BitstreamStorageService#clone. Is this what you had in mind?

mwoodiupui commented:
Yes, thank you.  Whenever we make changes to an entity which we do not commit, the caller should be warned.

marsaoua commented:
I tested it, a submitter can create a new version of its item!
Just a thought: Instead of changing the cloned bitstream in the BitstreamStorageServiceImpl#clone, does anything speak against moving all changes to BitstreamServiceImpl#clone?  In this way, there will not be a unpersisted state.

marsaoua commented:
We have created a pull request which avoid an unpersisted state in BitstreamStorageServiceImpl#clone
https://github.com/DSpace/DSpace/pull/2428#issue-279153056

tdonohue commented:
Closing, replaced by #2428 (Please review/test that PR so that we can get it resolved/merged)
